nixos/modules/services/monitoring/osquery/flags.nix

```diff
@@ -0,0 +1,36 @@
+{ cfg, lib, pkgs }:
+
+with builtins;
+with lib;
+rec {
+  # conf is the osquery configuration file used when the --config_plugin=filesystem.
+  # filesystem is the osquery default value for the config_plugin flag.
+  conf =
+    if usingFilesystemConfigPlugin then
```

Comment:
Another option: you could set `--config_plugin=filesystem` and `--config_path=${the derivation from the config file}` in the main `cfg.flags` unconditionally, *without* `lib.mkDefault`. This will require users to use `lib.mkForce` to override. That makes sense to me—any other plugins are definitely not the "happy path."

Reply:
> I think I'd prefer to make this conditional on cfg.config not being empty. rather than anybody manually setting --config_plugin=filesystem.

I think this is a nice shorthand and I was thinking it would also save us writing an empty derivation if there are no settings. However, if `--config_plugin=filesystem` the service will log an error `Error reading config: config file does not exist` which is not fatal but I think would be nice to avoid.

> So: if cfg.config isn't empty, we'd set --config_plugin and --config_path. We could consider a lib.warnIf when users try to set --config_plugin or --config_path (unless cfg.config is empty).

I'm not keen on the idea of warning about valid configuration. I do think it would be useful to note if a user is configures the `settings` option AND has explicitly set `--config_plugin` to something other than `filesystem`. That warning would let them know that their configuration is going to be ineffective.

Reply:
My priorities here:

1. `cfg.settings` *and* `cfg.flags` "just work" for most users.
2. If you do something "weird" that would result in us writing a config file but it getting ignored, you get a warning.
3. It should be clear from a quick skim of this file what's going on.

I think we're good on (1).

For (2), here's my understanding about the desired behaviors:

 | `--config_plugin`                 | `--config_path`                   | `cfg.settings` | User expectation                                       | What does the module do?                                                     |
| --- | --- | --- | --- | --- |
 | Empty (or `filesystem`)           | Empty                             | Empty          | N/A                                                    | Nothing                                                                      |
 | Empty (or `filesystem`)           | Empty                             | Nonempty       | `cfg.settings` works                                   | Writes `osquery.conf` to default path, and sets `--config_plugin=filesystem` |
 | Empty (or `filesystem`)           | OS path (`/var/lib/foo`)          | Empty          | They can write to `/tmp/foo` at runtime and it'll work | Nothing                                                                      |
 | Empty (or `filesystem`)           | OS path (`/var/lib/foo`)          | Nonempty       | Not sure.                                              | Warn that `cfg.settings` will be ignored                                     |
 | Empty (or `filesystem`)           | Derivation (`pkgs.writeText ...`) | Empty          | Use config from derivation                             | Nothing                                                                      |
 | Empty (or `filesystem`)           | Derivation (`pkgs.writeText ...`) | Nonempty       | Not sure.                                              | Warn that `cfg.settings` will be ignored                                     |
 | Something other than `filesystem` | Ignored                           | Empty          | Use that method to configure.                          | Nothing                                                                      |
 | Something other than `filesystem` | Ignored                           | Nonempty       | Not sure.                                              | Warn that `cfg.settings` will be ignored                                     |

If we agree on that, then the only question is how to make this code clear. My vote is probably:

- Remove the `if usingFilesystemConfigPlugin` check from `conf =`. We can rely on laziness not to write this if it's not needed.
- Set `flagfile = ... ({ config_path = conf; } // cfg.flags` just like it looks currently.
  - Add a `lib.warnIf` for `cfg.settings != {} && (runtimeValueIgnoringConf "config_plugin" "filesystem) != "filesystem") || cfg.flags ? "config_path")` (there might be a cleaner way to write this)

That's the main way we're using `runtimeValueIgnoringConf`. I almost think inlining `runtimeValueIgnoringConf` is clearer:

```nix
cfg.settings != {} && (
  (cfg.flags."config_plugin" or "filesystem") != "filesystem" ||
  cfg.flags ? "config_path"
)
```

Reply:
I resolve this discussion since @znewman01 approved the MR.